// angular-youtube-player/esm5/lib/ytplayer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { YTPlayerConfig } from './ytplayer.config';
import { loadAPI } from './util';
import * as i0 from "@angular/core";
import * as i1 from "./ytplayer.config";
var YTPlayerService = /** @class */ (function () {
    function YTPlayerService(config) {
        var _this = this;
        this.config = config;
        this.apiReady = new BehaviorSubject(window['YT'] !== undefined);
        this.players = new Array();
        this.multiplePlaying = config.multiplePlaying;
        if (config.shouldLoadAPI !== false) {
            loadAPI().then((/**
             * @return {?}
             */
            function () { return _this.apiReady.next(true); }));
        }
        else if (this.apiReady.value === false) {
            console.warn('YT API not loaded');
        }
    }
    Object.defineProperty(YTPlayerService.prototype, "playersCount", {
        get: /**
         * @return {?}
         */
        function () {
            return this.players.length;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} player
     * @return {?}
     */
    YTPlayerService.prototype.addPlayer = /**
     * @param {?} player
     * @return {?}
     */
    function (player) {
        this.players.push(player);
    };
    /**
     * @param {?} player
     * @return {?}
     */
    YTPlayerService.prototype.removePlayer = /**
     * @param {?} player
     * @return {?}
     */
    function (player) {
        /** @type {?} */
        var index = this.players.indexOf(player);
        this.players.splice(index, 1);
    };
    /**
     * @param {?} playerComponent
     * @return {?}
     */
    YTPlayerService.prototype.pauseAllExcept = /**
     * @param {?} playerComponent
     * @return {?}
     */
    function (playerComponent) {
        if (this.multiplePlaying) {
            return;
        }
        /** @type {?} */
        var exception = this.players.indexOf(playerComponent);
        if (exception > -1) {
            this.players.forEach((/**
             * @param {?} player
             * @param {?} index
             * @return {?}
             */
            function (player, index) {
                if (exception !== index) {
                    player.pause();
                }
            }));
        }
    };
    YTPlayerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    YTPlayerService.ctorParameters = function () { return [
        { type: YTPlayerConfig }
    ]; };
    /** @nocollapse */ YTPlayerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function YTPlayerService_Factory() { return new YTPlayerService(i0.ɵɵinject(i1.YTPlayerConfig)); }, token: YTPlayerService, providedIn: "root" });
    return YTPlayerService;
}());
export { YTPlayerService };
if (false) {
    /** @type {?} */
    YTPlayerService.prototype.apiReady;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.players;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.multiplePlaying;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.config;
}
