// angular-youtube-player/esm2015/lib/ytplayer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { YTPlayerConfig } from './ytplayer.config';
import { loadAPI } from './util';
import * as i0 from "@angular/core";
import * as i1 from "./ytplayer.config";
export class YTPlayerService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.apiReady = new BehaviorSubject(window['YT'] !== undefined);
        this.players = new Array();
        this.multiplePlaying = config.multiplePlaying;
        if (config.shouldLoadAPI !== false) {
            loadAPI().then((/**
             * @return {?}
             */
            () => this.apiReady.next(true)));
        }
        else if (this.apiReady.value === false) {
            console.warn('YT API not loaded');
        }
    }
    /**
     * @return {?}
     */
    get playersCount() {
        return this.players.length;
    }
    /**
     * @param {?} player
     * @return {?}
     */
    addPlayer(player) {
        this.players.push(player);
    }
    /**
     * @param {?} player
     * @return {?}
     */
    removePlayer(player) {
        /** @type {?} */
        const index = this.players.indexOf(player);
        this.players.splice(index, 1);
    }
    /**
     * @param {?} playerComponent
     * @return {?}
     */
    pauseAllExcept(playerComponent) {
        if (this.multiplePlaying) {
            return;
        }
        /** @type {?} */
        const exception = this.players.indexOf(playerComponent);
        if (exception > -1) {
            this.players.forEach((/**
             * @param {?} player
             * @param {?} index
             * @return {?}
             */
            (player, index) => {
                if (exception !== index) {
                    player.pause();
                }
            }));
        }
    }
}
YTPlayerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
YTPlayerService.ctorParameters = () => [
    { type: YTPlayerConfig }
];
/** @nocollapse */ YTPlayerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function YTPlayerService_Factory() { return new YTPlayerService(i0.ɵɵinject(i1.YTPlayerConfig)); }, token: YTPlayerService, providedIn: "root" });
if (false) {
    /** @type {?} */
    YTPlayerService.prototype.apiReady;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.players;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.multiplePlaying;
    /**
     * @type {?}
     * @private
     */
    YTPlayerService.prototype.config;
}
